x/foundation/keeper/grpc_query.go

```diff
@@ -224,3 +185,69 @@ func (s queryServer) TallyResult(c context.Context, req *foundation.QueryTallyRe
 
 	return &foundation.QueryTallyResultResponse{Tally: tally}, nil
 }
+
+func (s queryServer) Grants(c context.Context, req *foundation.QueryGrantsRequest) (*foundation.QueryGrantsResponse, error) {
+	if req == nil {
+		return nil, status.Error(codes.InvalidArgument, "invalid request")
+	}
+
+	if err := sdk.ValidateAccAddress(req.Grantee); err != nil {
+		return nil, err
+	}
+
+	ctx := sdk.UnwrapSDKContext(c)
+	store := ctx.KVStore(s.keeper.storeKey)
+	granteeKey := append(grantKeyPrefix, append([]byte{byte(len(req.Grantee))}, req.Grantee...)...)
+	grantStore := prefix.NewStore(store, granteeKey)
+
+	var authorizations []*codectypes.Any
+	if req.MsgTypeUrl != "" {
+		pageRes, err := query.Paginate(grantStore, req.Pagination, func(key []byte, value []byte) error {
+			var authorization foundation.Authorization
+			if err := s.keeper.cdc.UnmarshalInterface(value, &authorization); err != nil {
+				return err
+			}
+
+			msg, ok := authorization.(proto.Message)
+			if !ok {
+				return sdkerrors.ErrInvalidType.Wrapf("can't proto marshal %T", msg)
+			}
+			any, err := codectypes.NewAnyWithValue(msg)
+			if err != nil {
+				return err
+			}
+			authorizations = append(authorizations, any)
+
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		return &foundation.QueryGrantsResponse{Authorizations: authorizations, Pagination: pageRes}, nil
+	}
+
+	pageRes, err := query.Paginate(grantStore, req.Pagination, func(key []byte, value []byte) error {
+		var authorization foundation.Authorization
+		if err := s.keeper.cdc.UnmarshalInterface(value, &authorization); err != nil {
+			return err
+		}
+
+		msg, ok := authorization.(proto.Message)
+		if !ok {
+			return sdkerrors.ErrInvalidType.Wrapf("can't proto marshal %T", msg)
+		}
+		any, err := codectypes.NewAnyWithValue(msg)
+		if err != nil {
+			return err
+		}
+		authorizations = append(authorizations, any)
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &foundation.QueryGrantsResponse{Authorizations: authorizations, Pagination: pageRes}, nil
+}
```

Comment:
There is no logic that changes due to the empty of req.MsgTypeUrl. Why is it divided?

Reply:
It was my mistake. I updated the logic.